src/CSEjector.sol

```diff
@@ -0,0 +1,128 @@
+import { AccessControlEnumerableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/extensions/AccessControlEnumerableUpgradeable.sol";
+import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import { IStakingModule } from "./interfaces/IStakingModule.sol";
+import { ICSParametersRegistry } from "./interfaces/ICSParametersRegistry.sol";
+import { ICSAccounting } from "./interfaces/ICSAccounting.sol";
+import { ICSModule, NodeOperator } from "./interfaces/ICSModule.sol";
+import { ICSEjector } from "./interfaces/ICSEjector.sol";
+
+import { AssetRecoverer } from "./abstract/AssetRecoverer.sol";
+
+contract CSEjector is
+    ICSEjector,
+    Initializable,
+    AccessControlEnumerableUpgradeable,
+    PausableUntil,
+    AssetRecoverer
+{
+    bytes32 public constant PAUSE_ROLE = keccak256("PAUSE_ROLE");
+    bytes32 public constant RESUME_ROLE = keccak256("RESUME_ROLE");
+    bytes32 public constant BAD_PERFORMER_EJECTOR_ROLE =
+        keccak256("BAD_PERFORMER_EJECTOR_ROLE");
+    bytes32 public constant RECOVERER_ROLE = keccak256("RECOVERER_ROLE");
+
+    ICSModule public immutable MODULE;
+    ICSAccounting public immutable ACCOUNTING;
+
+    /// @dev see _keyPointer function for details of noKeyIndexPacked structure
+    mapping(uint256 noKeyIndexPacked => bool) private _isValidatorEjected;
+
+    constructor(address module) {
+        if (module == address(0)) revert ZeroModuleAddress();
+        MODULE = ICSModule(module);
+        ACCOUNTING = ICSAccounting(MODULE.accounting());
+    }
+
+    /// @notice initialize the contract from scratch
+    function initialize(address admin) external initializer {
+        if (admin == address(0)) revert ZeroAdminAddress();
+
+        __AccessControlEnumerable_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+    }
+
+    /// @inheritdoc ICSEjector
+    function resume() external onlyRole(RESUME_ROLE) {
+        _resume();
+    }
+
+    /// @inheritdoc ICSEjector
+    function pauseFor(uint256 duration) external onlyRole(PAUSE_ROLE) {
+        _pauseFor(duration);
+    }
+
+    /// @inheritdoc ICSEjector
+    function ejectBadPerformer(
+        uint256 nodeOperatorId,
+        uint256 keyIndex,
+        uint256 strikesCount
+    ) external whenResumed onlyRole(BAD_PERFORMER_EJECTOR_ROLE) {
+        _onlyExistingNodeOperator(nodeOperatorId);
+        NodeOperator memory no = MODULE.getNodeOperator(nodeOperatorId);
+        if (keyIndex >= no.totalDepositedKeys) {
+            revert SigningKeysInvalidOffset();
+        }
+
+        if (MODULE.isValidatorWithdrawn(nodeOperatorId, keyIndex))
+            revert AlreadyWithdrawn();
+
+        uint256 pointer = _keyPointer(nodeOperatorId, keyIndex);
+        if (_isValidatorEjected[pointer]) revert AlreadyEjected();
+
+        uint256 curveId = ACCOUNTING.getBondCurveId(nodeOperatorId);
+
+        ICSParametersRegistry registry = MODULE.PARAMETERS_REGISTRY();
```

Comment:
put this into immutable var

Reply:
Since ParametersRegistry is not in CSM v1, the approach proposed does not work with the current upgrade